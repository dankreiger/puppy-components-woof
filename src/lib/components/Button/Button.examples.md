Button examples:

###### default

```js
import { Button } from 'puppy-components-again';

<Button />;
```
