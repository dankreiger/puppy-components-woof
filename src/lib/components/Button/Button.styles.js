import styled from 'styled-components';

const buttonStyles = (
  color,
  colorHover,
  colorActive,
  bgColor,
  bgColorHover,
  bgColorActive,
  bgColorDisabled,
  disabled
) => `
  font-family: 'Prometo', Verdana, Arial, Helvetica, sans-serif;
  background-color: ${disabled ? bgColorDisabled : bgColor};
  font-weight: 500;
  color: ${color};
  border-radius: 30px;
  font-size: 16px;
  border: none;
  padding: 16px 24px 18px;
  transition: background-color 250ms;
  cursor: ${disabled ? 'not-allowed' : 'pointer'};
  &:hover {
    color: ${colorHover};
    background-color: ${disabled ? bgColorDisabled : bgColorHover};
    outline: none;
  }
  &:active {
    color: ${colorActive};
    background-color: ${disabled ? bgColorDisabled : bgColorActive};
    outline: none;
  }
  &:focus {
    outline: none;
  }
`;

export const ButtonStyled = styled.button`
  ${({ customStyles }) => customStyles};
  ${({
    color,
    colorHover,
    colorActive,
    bgColor,
    bgColorHover,
    bgColorActive,
    bgColorDisabled,
    disabled
  }) =>
    buttonStyles(
      color,
      colorHover,
      colorActive,
      bgColor,
      bgColorHover,
      bgColorActive,
      bgColorDisabled,
      disabled
    )}
`;

export const ButtonLinkStyled = styled.a`
  ${({
    color,
    colorHover,
    colorActive,
    bgColor,
    bgColorHover,
    bgColorActive,
    bgColorDisabled,
    disabled
  }) =>
    buttonStyles(
      color,
      colorHover,
      colorActive,
      bgColor,
      bgColorHover,
      bgColorActive,
      bgColorDisabled,
      disabled
    )}
  text-decoration: none;
  display: inline-block;
  ${({ customStyles }) => customStyles};
`;
