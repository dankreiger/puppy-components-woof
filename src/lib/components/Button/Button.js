import React from 'react';
import PropTypes from 'prop-types';
import { ButtonStyled, ButtonLinkStyled } from './Button.styles';
import {
  puppyCyanHover,
  puppyCyan,
  puppyBlue,
  puppyLightGray
} from '../../shared/colors';

const Button = ({
  className,
  customStyles,
  onClick,
  disabled,
  color,
  colorHover,
  colorActive,
  bgColor,
  bgColorHover,
  bgColorActive,
  bgColorDisabled,
  href,
  tag,
  target,
  text
}) => {
  /* html tag defined dynamically */
  if (tag === 'a') {
    return (
      <ButtonLinkStyled
        href={href}
        target={target}
        className={className}
        customStyles={customStyles}
        color={color}
        bgColor={bgColor}
        bgColorHover={bgColorHover}
        bgColorActive={bgColorActive}
        bgColorDisabled={bgColorDisabled}
        disabled={disabled}
      >
        {text}
      </ButtonLinkStyled>
    );
  } else {
    return (
      <ButtonStyled
        className={className}
        onClick={onClick}
        disabled={disabled}
        color={color}
        colorHover={colorHover}
        colorActive={colorActive}
        bgColor={bgColor}
        bgColorHover={bgColorHover}
        bgColorActive={bgColorActive}
        bgColorDisabled={bgColorDisabled}
        customStyles={customStyles}
      >
        {text}
      </ButtonStyled>
    );
  }
};

Button.propTypes = {
  className: PropTypes.string,
  customStyles: PropTypes.object,
  onClick: PropTypes.func,
  color: PropTypes.string,
  colorHover: PropTypes.string,
  colorActive: PropTypes.string,
  bgColor: PropTypes.string,
  bgColorHover: PropTypes.string,
  bgColorActive: PropTypes.string,
  bgColorDisabled: PropTypes.string,
  /** Disable the button */
  disabled: PropTypes.bool,
  /** href for links */
  href: PropTypes.string,
  /** Target for links (attribute or framename as custom string) */
  target: PropTypes.oneOf([
    '_blank',
    '_parent',
    '_top',
    '_self',
    PropTypes.string
  ]),
  /** HTML tag for button */
  tag: PropTypes.oneOf(['button', 'a']),
  /** Text of the button */
  text: PropTypes.string
};

Button.defaultProps = {
  onClick: e => e.preventDefault(),
  disabled: false,
  text: 'Click Me'
};

export default Button;
