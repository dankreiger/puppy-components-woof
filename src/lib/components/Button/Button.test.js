import React from 'react';
import { shallow } from 'enzyme';

import Button from './Button';

describe('<Button />', () => {
  describe('rendering', () => {
    it('should render', () => {
      const wrapper = shallow(<Button />);
      expect(wrapper).toBeDefined();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('defaults', () => {
    it('has the correct defaults', () => {
      expect(JSON.parse(JSON.stringify(Button.defaultProps))).toEqual({
        disabled: false,
        href: '#',
        tag: 'button',
        target: '_self',
        text: 'Click Me',
        theme: 'dark'
      });
    });
  });

  describe('button anchor tag', () => {
    it('should become an anchor tag when given an "a" tag prop', () => {
      const props = {
        tag: 'a'
      };
      const wrapper = shallow(<Button {...props} />);
      expect(wrapper.props().href).toBe('#');
      expect(wrapper.props().target).toBe('_self');
    });
  });

  describe('custom', () => {
    it('should allow custom className', () => {
      const props = {
        className: 'Custom'
      };
      const wrapper = shallow(<Button {...props} />);
      expect(wrapper.hasClass(props.className)).toBe(true);
    });
  });

  describe('behavior', () => {
    it('should call onClick', () => {
      const props = {
        onClick: jest.fn()
      };
      const wrapper = shallow(<Button {...props} />);
      wrapper.simulate('click');

      expect(props.onClick).toHaveBeenCalled();
    });
  });
});
