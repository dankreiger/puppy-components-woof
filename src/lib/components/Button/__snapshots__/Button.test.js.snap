// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> rendering should render 1`] = `ShallowWrapper {}`;
